### PR TITLE
feat: support previous special term when academic year changes

### LINKED ISSUE #4109: Do not override Special Term II courses when migrating academic year
### Is your feature request related to a problem? Please describe.

A clear and concise description of what the problem is. Eg. I'm always frustrated when [...]

### PR DESCRIPTION
**Pull request by jloh02**
Closes #4109

### CODE REVIEW

File: website/src/utils/specialTerm.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    return module;
+  }
+
+  if (
+    currentHasTimetable &&
+    archiveHasTimetable &&
+    JSON.stringify(currentSt2) === JSON.stringify(archiveSt2)
+  ) {
+    return module;
+  }
```
[the remark this answers is not in the document]
Comment (jloh02) [reply website/src/utils/specialTerm.ts:125]: Good catch — agreed. Expanded the JSDoc on `mergePreviousAySpecialTermData` (and extracted `shouldPreferArchiveSemesterData`) to document that archive data also wins when both timetables exist but differ, since ModReg R0 on the new AY is not authoritative during the overlap window. Fixed in cf81bde4.

File: scrapers/nus-v2/src/tasks/DataPipeline.ts [hunk 2]
```diff
@@ -45,19 +46,47 @@ export default class DataPipeline extends BaseTask implements Task<void, Array<M
...
+          ...semesterOrganizations,
+          modules: semesterModules,
         });
 
-        const { aliases } = await new CollateVenues(semester, this.academicYear).run(modules);
+        const { aliases } = await new CollateVenues(semester, semesterAcadYear).run(modules);
+
+        if (semester === 4 && usePreviousAySt2) {
+          await new CollateVenues(semester, this.academicYear).run(modules);
+        }
```
[the remark this answers is not in the document]
Comment (jloh02) [reply scrapers/nus-v2/src/tasks/DataPipeline.ts]: Agreed. Added an inline comment explaining that the second call writes venue files under the current AY for website lookups, and that its aliases are intentionally discarded because `CollateModules` uses the fetch-AY aliases. Fixed in cf81bde4.

File: website/src/actions/moduleBank.ts [hunk 3]
```diff
@@ -42,8 +44,44 @@ export const Internal = {
...
+
+  const mergedModule = mergePreviousAySt2Data(module, archiveModule);
+  if (mergedModule === module) {
+    return module;
+  }
+
+  dispatch({
+    type: SUCCESS_KEY(FETCH_MODULE),
+    payload: mergedModule,
+  });
```
[the remark this answers is not in the document]
Comment (jloh02) [reply website/src/actions/moduleBank.ts:84]: Partially agree on the trade-off, but this is intentional rather than accidental. `requestAction` must complete first so the module is in the store; archive enrichment is async and may require a separate fetch. Refactoring the request middleware to block the initial SUCCESS until enrichment finishes would be a much larger change for a narrow overlap window.

Also, ST II data is not necessarily *absent* in the gap — components may briefly show current-AY ModReg R0 placeholders before archive data replaces them, which is the case we are correcting.

Added a comment documenting the intentional re-dispatch in cf81bde4.

**Comment by ravern** [comment]
@greptileai review

File: packages/nusmods-academic-calendar/index.test.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  });
+
+  it('returns false before previous AY Special Term I starts', () => {
+    expect(isPreviousAySpecialTermActive('2025/2026', new Date(2025, 4, 1))).toBe(false);
+  });
+
+  it('returns false after new AY semester 1 starts', () => {
+    expect(isPreviousAySpecialTermActive('2025/2026', new Date(2025, 7, 11))).toBe(false);
+  });
+});
```
Comment (leslieyip02) [inline packages/nusmods-academic-calendar/index.test.js:32]: I think we should add tests for boundary values too:

```js
// AY 2024/2025 Special Term 1 begins on 2025-05-12
expect(isPreviousAySpecialTermActive('2025/2026', new Date(2025, 4, 12))).toBe(true);

// AY 2025/2026 Semester 1 beings on 2025-08-11
expect(isPreviousAySpecialTermActive('2025/2026', new Date(2025, 7, 10))).toBe(true);

// AY 2024/2025 Special Term 1 begins on 2025-05-12
expect(isPreviousAySpecialTermActive('2025/2026', new Date(2025, 4, 11))).toBe(false);
```

File: website/src/utils/specialTerm.test.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+vi.mock('config', () => ({
+  default: {
+    academicYear: '2025/2026',
+    specialTermAcademicYear: null,
+    showSt2ExamTimetable: true,
+  },
+}));
+
+describe(isPreviousAySpecialTermActive, () => {
```
Comment (leslieyip02) [inline website/src/utils/specialTerm.test.ts:23]: Do we need to dupe these tests? They should be covered by `packages/nusmods-academic-calendar/index.test.js`.

**Review by leslieyip02 — CHANGES_REQUESTED** [review]
Mostly LGTM, just some minor nits.

**Review by leslieyip02 — APPROVED** [review]
